android/app/src/main/java/org/mozilla/magnet/magnetapi/ApiPreferences.java

```diff
@@ -0,0 +1,106 @@
+package org.mozilla.magnet.magnetapi;
+
+import android.content.Context;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.mozilla.magnet.api.Api;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.Map;
+import java.util.HashMap;
+
+class ApiPreferences extends Api {
+
+    private static final String PREFERENCES_STORE_KEY = "preferences";
+    private static final String PREFERENCES_KEY = "pref_key";
+
+    private final Lock mWriteLock = new ReentrantLock();
+
+    ApiPreferences(Context aContext) {
+        super(aContext);
+    }
+
+    /**
+     * Get user's preferences.
+     *
+     */
+    @Override
+    public void get(String aPath, Callback aCallback) {
+        aCallback.resolve(getCache().getJsonObject(PREFERENCES_STORE_KEY));
+    }
+
+    /**
+     * Update or set a preference given a Map with structure:
+     * `{ "pref_key": "string", "value": "string" }`
+     */
+    @Override
+    public void post(String aPath, Object aData, Callback aCallback) {
+        if (!(aData instanceof Map)) {
+            aCallback.reject("Data must be a Map type");
+            return;
+        }
+
+        // Lock everything between getting the JSON object from the cache and
+        // writing it back - a synchronize on the method isn't enough as the
+        // resource is used in the delete method too.
+        mWriteLock.lock();
```

Comment:
Unfortunately in Java you can't scope a lock, there isn't, as far as I know, a way to do RAII (Resource acquisition is initialization) to scope the lock and have it release automatically. Although it looks like there might be a way to do this in Java 7, I'm not sure if it works with Locks, I should try.

So given this, the code between calling 'lock' and 'unlock' is the scope, aka the critical section.

A lock is a mutual exclusion mechanism, it ensures that only one thread is able to execute the code between acquiring the lock, and releasing the lock.  The code block protected by a lock is called the critical section.  If a thread "holds the lock", no other thread can enter the critical section until it acquires the lock-when the other thread has releases the lock.  It provides a way of linearising access to a resource and ensuring certain operations on shared memory are atomic.  An example directly applicable here is, we don't want to add an entry, and then read it back to find that during the add operation, another thread actually deleted it.

Java has the 'synchronised' keyword that will mutex lock a single method.  But we need to lock across a couple of methods here.